### DISCUSSION TITLE
Bridging Swift Code

### DISCUSSION THREAD

**Discussion by bitigchi** [transmission/transmission#4744] [Q&A]
I am trying to add some piece of Swift code in a separate file, but although I have created the bridging header, I am unable to import it via:

```swift
#import "Transmission-Swift.h"
```

Probably I need to adjust the `CMakesLists.txt`, but I am totally clueless on these things. Any ideas?

**Comment by ckerr** [comment]
CC @Coeur

**Reply by Coeur** [reply]
Hi @bitigchi.

See the two commits of this PR: https://github.com/transmission/transmission/pull/4540
In particular, the first commit is roughly the minimal changes needed for supporting Swift, while the second commit is actual Objective-C to Swift conversion for various project files.

When using CMakeLists.txt, you have various ways to build:
1. `cmake -B build` + `make -C build`
2. `cmake -B build -G Ninja` + `ninja -C build transmission-mac`
3. `cmake -B build -G Xcode` + build with Xcode or `xcrun` or `xcodebuild`

But:
1. `make` isn't compatible with Swift at all.
2. `ninja` is theoretically compatible with Swift, but I was unsuccessful at it so far.
3. `Xcode` solution works well.

**Reply by bitigchi** [reply]
Thanks for the detailed reply. I added my file and the bridging header to `macosx/CMakeLists.txt`, but still couldn't get it to work with the minimal `../CMakeLists.txt` changes in the first commit.

Would it be possible to try a basic `#import` mentioned in my original message? Just to rule out any error from my part.

BTW, that PR is awesome, really making things easier.

**Reply by Coeur** [reply]
There is one `#import <Transmission-Swift.h>` in Transmission+CXX.h
And you'll need to build with Xcode to get it to work.

**Reply by Coeur** [reply]
Two months that my PR is open (#4540), and I finally got someone to look at my "help needed" for cmake Today (through my PR which only takes the first commit, #4757). ;)
Apparently, `XCODE_ATTRIBUTE_*` (like XCODE_ATTRIBUTE_CLANG_ENABLE_MODULES) are only used with `cmake -G Xcode` but not `cmake -G Ninja`, which could explain why Ninja builds are failing? I'll try to confirm if that's the case or not, and search for the Ninja equivalent to enable module and bridging header.

**Reply by ckerr** [reply]
> Two months that my PR is open (https://github.com/transmission/transmission/pull/4540), and I finally got someone to look at my "help needed" for cmake Today

It's a +7,363 / -7,471 PR refactor that you introduced during a beta cycle! :joy_cat: 

@Coeur I love you but of course a 14k line refactor PR is gonna take a backseat during beta & bugfix  :green_heart:

**Reply by Coeur** [reply]
It's not to debug all those lines. Just the CMakeLists.txt to support Ninja (CMakeLists.txt works for Xcode).
What surprises me is that no one else has this issue on internet:
https://www.google.com/search?q=%22ninja%22+%22fatal+error:%22+%22Swift.h%22+%22file+not+found%22
[edit]
Actually one other person in the world had this issue, Facebook, in 2017, and they never managed to solve it (instead, they abandoned the project): https://github.com/facebookarchive/xcbuild/issues/237

So maybe no one uses a CMakeLists to build a mixed ObjC and Swift project using Ninja. (pure Swift would work, or using Xcode exclusively would work)

**Reply by Coeur** [reply]
I've tried all I could, I've ran out of ideas to fix the ninja build. It seems that ninja never builds any Swift.h file, so that "Transmission-Swift.h" doesn't exist and build fails. Not sure if it needs to be addressed by cmake team or ninja team.

For me, it means that currently in a same target I can use ObjC code in Swift, but not Swift code in ObjC.
We could try to:
- either split the code in two targets (one ObjC, one Swift, and each one exposes an API to the other)
- or convert all the ObjC code to Swift (then no need to use Swift in ObjC anymore)
- or drop support for Ninja for the macOS app (we can still build from command line with `xcodebuild -project build/transmission.xcodeproj -target transmission-mac`, so CI and everybody can still build)
- or we get surprise solution for fixing the build with Ninja (but after so much time on it, it may be impossible)

**Reply by GaryElshaw** [reply]
Sorry it's been such a nightmare, Antoine.

**Reply by Coeur** [reply]
The "easiest" solution for me is:
> convert all the ObjC code to Swift (then no need to use Swift in ObjC anymore)

But then, instead of a +7,363 / -7,471 PR, it may be closer to a +20,000 / -20,000 PR.

**Reply by GaryElshaw** [reply]
What's your feeling about that? Are there choices there that you think are major blockers or drawbacks? Size, memory consumption, the personal time and energy?

**Reply by Coeur** [reply]
Performances should be identical (if not better) in my opinion.
Personal time and energy is the biggest drawback: I may not complete that second half of the work before a year from now.

Actually, "easiest" is to drop Ninja support and only rely on Xcodebuild, but it's not ideal.

**Reply by Coeur** [reply]
Ah, I think I found an easier workaround: @bitigchi we don't need that autogenerated "Transmission-Swift.h" file. We can simply copy-paste its content in a hardcoded file. I know that some projects are doing this. I'll try it when I have the time.

**Reply by bitigchi** [reply]
Waiting for it!

**Comment by GaryElshaw** [comment]
You have an excellent use-case for [ninja](https://github.com/ninja-build/ninja), i'd think they would be interested. Xcode and cmake+make if they don't come to the party? That's still pretty good insofar as choices go.

Time. Are any of the following tools useful? https://www.macupdate.com/find/mac/context%3Dswift%2520converter

**Reply by Coeur** [reply]
I've just opened an issue for ninja, let's see what they say about it:
https://github.com/ninja-build/ninja/issues/2261

As for apps that convert code, I have my doubts: we're not in ObjC, we're in ObjC++... and also most the real work is to deal manually with what can't be automated.
